sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala

```diff
@@ -26,34 +26,39 @@ abstract class FileScanBuilder(
     fileIndex: PartitioningAwareFileIndex,
     dataSchema: StructType) extends ScanBuilder with SupportsPushDownRequiredColumns {
   private val partitionSchema = fileIndex.partitionSchema
+  private val partitionNameSet = toNameSet(partitionSchema)
   private val isCaseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
   protected var requiredSchema = StructType(dataSchema.fields ++ partitionSchema.fields)
 
   override def pruneColumns(requiredSchema: StructType): Unit = {
-    this.requiredSchema = requiredSchema
+    // the default implementation prunes only top-level columns
+    // file formats that support nested column pruning must override this
+    val fields = dataSchema.fields ++ partitionSchema.fields
+    val requiredNameSet = toNameSet(requiredSchema)
+    val requiredFields = fields.filter { field =>
+      val colName = PartitioningUtils.getColName(field, isCaseSensitive)
+      requiredNameSet.contains(colName)
+    }
+    this.requiredSchema = StructType(requiredFields)
   }
 
   protected def readDataSchema(): StructType = {
-    val requiredNameSet = createRequiredNameSet()
-    val fields = dataSchema.fields.filter { field =>
+    val dataFields = requiredSchema.fields.filter { field =>
```

Comment:
We iterate through `requiredSchema` to pick changes to nested columns (i.e. `dataSchema` can contain the old struct).